src/main/java/org/apache/sysds/runtime/instructions/fed/UnaryMatrixFEDInstruction.java

```diff
@@ -63,17 +72,316 @@ public static UnaryMatrixFEDInstruction parseInstruction(String str) {
+		String tmpInstString = instString.replace("ucumk+*", "ucum*");
+
+		FederatedRequest fr1 = FederationUtils.callInstruction(tmpInstString, output,
+			new CPOperand[] {input1}, new long[] {mo1.getFedMapping().getID()});
+		FederatedRequest fr2 = new FederatedRequest(FederatedRequest.RequestType.GET_VAR, fr1.getID());
+		Future<FederatedResponse>[] tmp = mo1.getFedMapping().execute(getTID(), true, fr1, fr2);
+
+		// slice and construct prod matrix
+		MatrixBlock ret = new MatrixBlock(tmp.length, 2, 0.0);
+		for(int i = 0; i < tmp.length; i++)
+			try {
+				MatrixBlock curr = ((MatrixBlock) tmp[i].get().getData()[0]);
+				ret.setValue(i, 1, curr.getValue(curr.getNumRows()-1, 1));
+			}
+			catch(Exception e) {
+				throw new DMLRuntimeException("Federated Get data failed with exception on UnaryMatrixFEDInstruction",
+					e);
+			}
+		return ret;
+	}
+
+	private NewVariable[] getTernaryVars(ExecutionContext ec, MatrixObject mo1, MatrixBlock scalingValues) {
+		MatrixBlock condition = new MatrixBlock((int) mo1.getNumRows(), (int) mo1.getNumColumns(), 1.0);
+		MatrixBlock mb2 = new MatrixBlock((int) mo1.getNumRows(), (int) mo1.getNumColumns(), 0.0);
+
+		int step;
+		for(int i = 0; i < scalingValues.getNumRows()-1; i++) {
+			step = (int) mo1.getFedMapping().getFederatedRanges()[i+1].getBeginDims()[0];
+			condition.setValue(step, 0, 0.0);
+			mb2.setValue(step, 0, scalingValues.getValue(i, 0));
+		}
+
+		MatrixObject cond = ExecutionContext.createMatrixObject(condition);
+		long condID = FederationUtils.getNextFedDataID();
+		ec.setVariable(String.valueOf(condID), cond);
+
+		MatrixObject mo2 = ExecutionContext.createMatrixObject(mb2);
+		long varID2 = FederationUtils.getNextFedDataID();
+		ec.setVariable(String.valueOf(varID2), mo2);
+
+		return new NewVariable[] {new NewVariable(condID, cond), new NewVariable(varID2, mo2)};
+	}
+
+	private void setScalingValues(ExecutionContext ec, MatrixObject mo1, MatrixObject out, NewVariable[] ternaryVars) {
+		CPOperand opCond = new CPOperand(String.valueOf(ternaryVars[0]._id), ValueType.FP64, DataType.MATRIX);
+		CPOperand op2 = new CPOperand(String.valueOf(ternaryVars[1]._id), ValueType.FP64, DataType.MATRIX);
+
+		String ternaryInstString = constructTernaryString(opCond, op2);
+		FederatedRequest[] fr1 = mo1.getFedMapping().broadcastSliced(ternaryVars[0]._mo, false);
+		FederatedRequest[] fr2 = mo1.getFedMapping().broadcastSliced(ternaryVars[1]._mo, false);
+		FederatedRequest fr3 = FederationUtils.callInstruction(ternaryInstString, output,
+			new CPOperand[] {input1, opCond, op2}, new long[] {mo1.getFedMapping().getID(), fr1[0].getID(), fr2[0].getID()});
+		mo1.getFedMapping().execute(getTID(), true, fr1, fr2, fr3);
+
+		out.setFedMapping(mo1.getFedMapping().copyWithNewID(fr3.getID()));
+
+		ec.removeVariable(opCond.getName());
+		ec.removeVariable(op2.getName());
+	}
+
+	private void aggSumprod(MatrixObject out) {
+		instString = setOutputVarInInstString(output);
```

Comment:
please don't modify the instString which is a member variable used potentially many times when executing this instruction in a loop.